### PR TITLE
Add Shopify Order Line Attributes

### PR DESCRIPTION
**Pull request by louagej**
customAttributes are already in the GraphQL query but aren't saved are displayed anywhere.
We have some business Case where Shopify Adds customAttributes to an order line.
This proposal is to save and display these key values in a same manner as already done with the customAttributes for the Shpfy Order Header.

### CODE REVIEW

File: Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        field(1; "Order Id"; BigInteger)
+        {
+            Caption = 'Order Id';
+            DataClassification = CustomerContent;
+        }
+        field(2; "Order Line Id"; Guid)
+        {
+            Caption = 'Line Id';
+            DataClassification = CustomerContent;
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:19]: I think these fields should habe `DataClassification = SystemMetaData;`

File: Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/// <summary>
+/// Table Shpfy Order Line Attribute (ID 30123).
+/// </summary>
+table 30148 "Shpfy Order Line Attribute"
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:4]: The default value for Shopify Objects is Access = Internal and extensible = false.
If you need some table public, then please outline the scenario for discussion with microsoft.

File: Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/// <summary>
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:1]: All new Objects require the OSS Header (see other objects)

File: Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al [hunk 4]
```diff
@@ -299,6 +299,25 @@ codeunit 30161 "Shpfy Import Order"
...
     end;
 
+    [NonDebuggable]
+    local procedure ImportCustomAttributtes(ShopifyOrderId: BigInteger; OrderLineId: Guid; JCustomAttributtes: JsonArray)
+    var
+        OrderLineAttribute: Record "Shpfy Order Line Attribute";
+        JToken: JsonToken;
+    begin
+        OrderLineAttribute.SetRange("Order Id", ShopifyOrderId);
+        if not OrderLineAttribute.IsEmpty then
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al:309]: IsEmpty is procedure and should have paranetheses.

File: Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al [hunk 5]
```diff
@@ -299,6 +299,25 @@ codeunit 30161 "Shpfy Import Order"
...
+            OrderLineAttribute.DeleteAll();
+        foreach JToken in JCustomAttributtes do begin
+            Clear(OrderLineAttribute);
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al:312]: I don't know how microsoft is working with this, but we in our company use always init and Insert(true), despite in for special scenarios (temporary tables..)

File: Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al [hunk 6]
```diff
@@ -299,6 +299,25 @@ codeunit 30161 "Shpfy Import Order"
...
+            OrderLineAttribute."Order Id" := ShopifyOrderId;
+            OrderLineAttribute."Order Line Id" := OrderLineId;
+            OrderLineAttribute.Key := JsonHelper.GetValueAsText(JToken, 'key', MaxStrLen(OrderLineAttribute."Key"));
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al:315]: I think the compiler will still warn although the value isn't overflowing the field.

File: Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            Caption = 'Order Id';
+            DataClassification = CustomerContent;
+        }
+        field(2; "Order Line Id"; Guid)
+        {
+            Caption = 'Line Id';
+            DataClassification = CustomerContent;
+        }
+        field(3; "Key"; Text[100])
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:21]: Is there any documentation about the key length for custom attributes?
I'm unable to find any documentation on this topic for shopify.

Comment (louagej) [reply Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:21]: In Shopify it is documented as a string, https://shopify.dev/docs/api/storefront/2023-07/input-objects/AttributeInput
That's the reason why I used Copystr to fill the value in this field

File: Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al (hunk 1 above)
Comment (louagej) [reply Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:19]: ok, Done

File: Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al (hunk 2 above)
Comment (louagej) [reply Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:4]: ok, added properties

File: Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al (hunk 3 above)
Comment (louagej) [reply Apps/W1/Shopify/app/src/Order handling/Tables/ShpfyOrderLineAttribute.Table.al:1]: ok, added OSS Header

File: Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al (hunk 4 above)
Comment (louagej) [reply Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al:309]: done

File: Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al (hunk 6 above)
Comment (louagej) [reply Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al:315]: I agree

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

Build ID: 12060.

**Pull request closed without merging**